### PR TITLE
[SDP-1025]: New endpoint to cancel individual payment in ready status

### PR DESCRIPTION
**Pull request by ceciliaromao**
### What

Implement a new endpoint PATCH /payments/{id}/status that allows the user to change the status of an individual payment

### Why

 UNHCR asked for the ability to cancel individual payments. 

### Known limitations

Make sure to return an error if the transition is not possible. We only allow payments in ready status to be cancelled. There is a state machine that can be used to check if the transition is possible.

#### PR Structure

* [x] This PR has a reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR title and description are clear enough for anyone to review it.
* [x] This PR does not mix refactoring changes with feature changes (split into two PRs otherwise).

#### Thoroughness

* [x] This PR adds tests for the new functionality or fixes.
* [x] This PR contains the link to the Jira ticket it addresses.

#### Configs and Secrets

* [x] No new CONFIG variables are required -OR- the new required ones were added to the helmchart's [`values.yaml`] file.
* [x] No new CONFIG variables are required -OR- the new required ones were added to the deployments ([`pr-preview`], [`dev`], [`demo`], `prd`).
* [x] No new SECRETS variables are required -OR- the new required ones were mentioned in the helmchart's [`values.yaml`] file.
* [x] No new SECRETS variables are required -OR- the new required ones were added to the deployments ([`pr-preview secrets`], [`dev secrets`], [`demo secrets`], `prd secrets`).

#### Release

* [x] This is not a breaking change.
* [x] **This is ready for production.**. If your PR is not ready for production, please consider opening additional complementary PRs using this one as the base. Only merge this into `develop` or `main` after it's ready for production!

#### Deployment

* [ ] Does the deployment work after merging?

[`values.yaml`]: ../helmchart/sdp/values.yaml
[`pr-preview`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/common-previews/stellar-disbursement-platform/backend-helm-values
[`dev`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/backend-helm-values
[`demo`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/demo/demo-backend-helm-values
[`pr-preview secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/common-previews/externalsecrets-common-previews.yaml#L241-L346
[`dev secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/stellar-disbursement-platform-externalsecrets.yaml
[`demo secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/demo/demo-sdp-externalsecrets.yaml

### CODE REVIEW

File: internal/services/payment_management_service.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+// CancelPayment update payment to status 'canceled'
+func (s *PaymentManagementService) CancelPayment(ctx context.Context, paymentID string) error {
+	return db.RunInTransaction(ctx, s.dbConnectionPool, nil, func(dbTx db.DBTransaction) error {
+		payment, err := s.models.Payment.Get(ctx, paymentID, dbTx)
+		if err != nil {
+			if errors.Is(err, data.ErrRecordNotFound) {
+				return ErrPaymentNotFound
+			} else {
+				return fmt.Errorf("error getting disbursement with id %s: %w", paymentID, err)
```
Comment (marwen-abid) [inline internal/services/payment_management_service.go:43]:
```suggestion
				return fmt.Errorf("error getting payment with id %s: %w", paymentID, err)
```

**Review by marwen-abid — APPROVED** [review]
LGTM! Few minor comments but this looks great!

File: internal/serve/httphandler/payments_handler.go [hunk 2]
```diff
@@ -153,3 +155,58 @@ func (p PaymentsHandler) getPaymentsWithCount(ctx context.Context, queryParams *
...
+	ctx := r.Context()
+	paymentID := chi.URLParam(r, "id")
+
+	switch toStatus {
+	case data.CanceledPaymentStatus:
+		err = paymentManagementService.CancelPayment(ctx, paymentID)
+		response.Message = "Payment canceled"
+	default:
+		err = services.ErrPaymentStatusCantBeChanged
+	}
```
Comment (CaioTeixeira95) [inline internal/serve/httphandler/payments_handler.go:194]: Why not just use early return here?

```suggestion
	if toStatus != data.CanceledPaymentStatus {
		httperror.BadRequest(services.ErrPaymentStatusCantBeChanged.Error(), err, nil).Render(w)
		return
	}
	
	err = paymentManagementService.CancelPayment(ctx, paymentID)
	if err != nil {
		...
	}
```

**Review by CaioTeixeira95 — APPROVED** [review]
LGTM 🚀 

Thank you for adding tests for other functions. 🥇 I've left some suggestions. There is no need to implement them if you don't think it's necessary.

File: internal/serve/httphandler/payments_handler.go (hunk 2 above)
Comment (ceciliaromao) [reply internal/serve/httphandler/payments_handler.go:194]: I used `switch` to patch other statuses if it's implemented in the future...

**Pull request merged**